### ISSUE TITLE
Autogenerate does not consider schema for enum type

### ISSUE THREAD

**Issue by renedupont** [sqlalchemy/alembic#899]
**Describe the bug**
I am using Postgres 12, Alembic (1.7.1) and SQLAlchemy (1.4.23) 
and have something like this SQLAlchemy enum in my model:
```
class MyEnum(enum.Enum):
    one= 'one'
    two= 'two'

my_enum = Column(Enum(MyEnum), nullable=False)
```

When I run
`alembic revision --autogenerate`

I get in my migration script:
```
sa.Column('my_enum',
          sa.Enum('one', 'two', name='myenum'),
          nullable=False),
```

I am using another schema than the default (public) one and while there is a schema arg that can be set (and I am currently changing the script manually adding it), it is not being considered by autogenerate. For stuff like tables and so on, it works perfectly and the schema is considered.

**Expected behavior**
I'd like to see the following being autogenerated for my example case:
```
sa.Column('my_enum',
          sa.Enum('one', 'two', name='myenum', schema='my_schema_name'), 
          nullable=False),
```

**Versions.**
 - OS: Ubuntu 20.04
 - Python: 3.8.10
 - Alembic: 1.7.1
 - SQLAlchemy: 1.4.23
 - Database: Postgres 12
 - DBAPI:

**Comment by CaselIT** [comment]
Thanks for reporting.

>I am using another schema than the default (public) one and while there is a schema arg that can be se

is the schema set on the MetaData?

**Comment by renedupont** [comment]
In `env.py` I have
```
from myapp.models import Base
... 
target_metadata = Base.metadata
```
and in myapp.models I do:
```
engine: Engine = create_engine(myDbUrl, pool_pre_ping=True)
db_session: scoped_session = scoped_session(sessionmaker(autocommit=False, autoflush=False, bind=engine))
Base = declarative_base()
Base.query = db_session.query_property()
```

**Comment by CaselIT** [comment]
Where is the schema defined?

**Comment by renedupont** [comment]
In `myapp.models` I have a base model like this where I set the schema:
```
class BaseModel(Base):
    __abstract__ = True
    __table_args__ = {'extend_existing': True, 'schema': 'my_schema_name'}
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
```
The model class that contains the enum in question is derived from this base model class.

**Comment by renedupont** [comment]
Looking at it again, I think I have to add more info of the scenario described in my opening post. The auto-generated full example looks actually more like this:
```
    op.create_table('my_table_name',
                              sa.Column('id', postgresql.UUID(as_uuid=True), nullable=False),
                              sa.Column('my_enum',
                                                sa.Enum('one', 'two', name='myenum'), nullable=False),
                              sa.PrimaryKeyConstraint('id'),
                              schema='my_schema_name'
                              )
```
So it has a schema at the create_table level, but when running this script, the enum which gets created as a postgres TYPE in the database is still being created in the default (public) schema. When adding the schema name to the sa.Enum(...) as mentioned in my opening post in the expected behaviour part, it is being created in the right schema.

I hope that helped explaining a bit more in case it was still confusing.

**Comment by vinayman** [comment]
Is there any more news on this request? I am encountering the exact same issue

**Comment by zzzeek** [comment]
no.  please edit your files manually

**Comment by smiling-watermelon** [comment]
@zzzeek, by:
> no. please edit your files manually

Do you mean that the fix for this will not be created, or do you mean that at the moment there is no fix, and for now, we should make manual adjustments, and in the future this will be fixed?

---
I have an even comical situation in this regard:
```python
...
    op.alter_column(
        "entity_name"
        "status",
        existing_type=sa.Enum("INITIAL", "PENDING", "ERROR", "SUCCESS", name="status"),
        type_=postgresql.ENUM(
            "INITIAL",
            "PENDING",
            "ERROR",
            "SUCCESS",
            name="status",
            schema="schema_name",
        ),
        existing_nullable=False,
        schema="schema_name",
    )
...
```

And migration fails like this:
```
sqlalchemy.exc.ProgrammingError: (psycopg.errors.UndefinedObject) type "statustype" does not exist
[SQL: ALTER TABLE schema_name.entity_name ALTER COLUMN type TYPE statustype ]
(Background on this error at: https://sqlalche.me/e/20/f405)
```

So it doesn't see the schema of a `sa.Enum` object (rightfully so - it is not set, even though in my model I have set it for `ENUM` object explicitly), even though `postgresql.ENUM` does have schema specified.
Or so I presume :cry:

**Comment by zzzeek** [comment]
> @zzzeek, by:
> 
> > no. please edit your files manually
> 
> Do you mean that the fix for this will not be created, or do you mean that at the moment there is no fix, and for now, we should make manual adjustments, and in the future this will be fixed?

the bug is opened, which means to the degree that maintenance on Alembic continues, this bug is part of what's considered available work.

that said, ENUM is a wide, sprawling issue for which I maintain only a holistic solution would be appropriate, that is, major architectural / workflow enhancements to Alembic to support this single, PostgreSQL-only datatype, so I have no plans in the near term to work on it (this has been the case for some years now)

**Comment by obrazkowv** [comment]
Hi!

Are there any plans to solve this problem. Working with Enum type when there is a custom schema becomes much more complicated because alembic thinks that column type has changed. And i need  to manually remove all false positive update on each migration generation.

Maybe you see a workaround to work around this?

Thanks in advance

**Comment by zzzeek** [comment]
@CaselIT how do the upcoming changes in enum schema affect this ?   I guess rendering the schema explicitly is still the same solution correct ?   

There is an overall architectural issue regarding how create type, drop type, and most importantly alter type, should be done for a PG enum.   But this issue is just rendering explicit schema name that matches what the effective schema of this enum should be.   This...should be a small fix ?  

Overall I would think the workaround would be to use inherit_schema=True on the enum.   That should autogenerate.   Then it will use the same schema as that of the table.

**Comment by obrazkowv** [comment]
Alembic generates correct migration on first generation. But next time alembic tries to alter_column to update type because existing type does not include schema. inherit_schema=True  didn't help me to resolve this problem
```
metadata = MetaData(schema="some_schema")
Table(
    "sometable",
    metadata,
    Column("status", ENUM(Status, schema=metadata.schema), nullable=False),
)
```

Alembic produces this on first migration
```
op.create_table(
    "some_table",
    sa.Column("status", sa.Enum( "PENDING", "FINISHED", name="somestatus", schema="some_schema"), nullable=False),
    schema="some_schema",
)
```

Alembic produces this on subsequent migrations
```
op.alter_column(
    'some_table', 'status',
    existing_type=postgresql.ENUM('PENDING', 'FINISHED', name='somestatus'),
    type_=postgresql.ENUM('PENDING', 'FINISHED', name='somestatus', schema='some_schema'),
    existing_nullable=False,
    schema='some_schema'
)
```

**Comment by zzzeek** [comment]
How about set the schema on the enum explicitly ?

**Comment by obrazkowv** [comment]
That what i actually do and this is my pain at the same time:) Each new migration generates alter column on all enums that have explicitly specified schema. And i hope one day this will be fixed, if this problem is still not clear, i'll try to provide more context.
Thanks for your time!

**Comment by zzzeek** [comment]
> That what i actually do and this is my pain at the same time:) Each new migration generates alter column on all enums that have explicitly specified schema. And i hope one day this will be fixed, if this problem is still not clear, i'll try to provide more context. Thanks for your time!

do you have schema set on MetaData?   While we can get the schema to render here, in SQLAlchemy 2.1 an Enum will no longer inherit the schema from the parent Table, inherit_schema is still there but is being deprecated.

if your  MetaData does not have the schema, then putting the explicit schema on the Enum is the way to go anyway.  these objects live in the Postgresql database independently of tables that use them.

**Comment by obrazkowv** [comment]
Yes, my metadata knows about schema. Schema used for enums and tables is the same but it differs from "public" and because i don't use postgres "search path" i have to specify schema on different objects.

If you look at this comment you'll see what i actually do. First block code - definition of the Table with MetaData that knows about schema. This table contains column with Enum type which directly associated with "some_schema". So, migration ,as expected, creates table and enum in "some_schema" schema. And this is totally fine after this you can upgrade and on next autogenerate migration will produce third block of code with "alter_column" operation
https://github.com/sqlalchemy/alembic/issues/899#issuecomment-3472816343

**Comment by CaselIT** [comment]
Does it happen also if you start from an empty db and create all tables with create_all? If tou then run alembic there does it report any change?

**Comment by obrazkowv** [comment]
No, it happens only on generation new migration after appliying migration with Enum type column.
Steps to reproduce:
1. Create table with Enum column or add Enum column to the existing table. Enum must have non-nullable schema(specified) schema to your table(add new column and autogenerate migration)
2. Autogenerate + upgrade alembic migration
3. Run autogenerate one more time and you will see changes related to migration that you've upgraded on step 3. This migration is created because 

https://github.com/sqlalchemy/alembic/blob/main/alembic/autogenerate/compare.py#L406
I can be mistaken but I think conn_col variable does not know about database "schema"(current state of database) while metadata_col knows about schema (current state of application) and this diff triggers new migration

**Comment by CaselIT** [comment]
I cannot reproduce what you see.

- I started from an empty db
- created `test_schema`
- created a new alembic env with `alembic init -t pyproject .`
- added a file like so
    ```py
    import sqlalchemy as sa

    meta = sa.MetaData()

    t = sa.Table(
        "t1",
        meta,
        sa.Column("c1", sa.Enum("x", "y", name="e1", schema="test_schema")),
        schema='test_schema'
    )
    ```
- updated `env.py` to set `target_metadata = meta` and I added `include_schemas=True,` to `context.configure` in `run_migrations_online`
- I created a fist migration with `alembic revision --autogenerate -m 'one'`:
  ```py
    """one

    Revision ID: 5b693ddef1c2
    Revises: 
    Create Date: 2025-11-04 21:39:20.332952

    """
    from typing import Sequence, Union

    from alembic import op
    import sqlalchemy as sa


    # revision identifiers, used by Alembic.
    revision: str = '5b693ddef1c2'
    down_revision: Union[str, Sequence[str], None] = None
    branch_labels: Union[str, Sequence[str], None] = None
    depends_on: Union[str, Sequence[str], None] = None


    def upgrade() -> None:
        """Upgrade schema."""
        op.create_table('t1',
        sa.Column('c1', sa.Enum('x', 'y', name='e1', schema='test_schema'), nullable=True),
        schema='test_schema'
        )

    def downgrade() -> None:
        """Downgrade schema."""
        op.drop_table('t1', schema='test_schema')
  ```
- applied it to the db by running `alembic upgrade heads`
- now if I run `alembic revision --autogenerate -m 'two'` no changes is detected and an empty migration is generated


could you provide the step you are using to trigger the issue? thanks!

**Comment by obrazkowv** [comment]
I'll try to find a reason why we see this difference in behaviour.
Tnx for your time and help i appreciate it.

**Comment by obrazkowv** [comment]
I found the difference, i set search_path in alembic env file. I'm not sure that it works as expected, what do you think about that?
```
def do_run_migrations(connection: Connection) -> None:
    def process_revision_directives(
        _: MigrationContext,
        revision: str | Iterable[str | None] | Iterable[str],
        directives: list[MigrationScript],
    ):
        assert config.cmd_opts is not None
        if getattr(config.cmd_opts, "autogenerate", False):
            script = directives[0]
            assert script.upgrade_ops is not None
            if script.upgrade_ops.is_empty():
                directives[:] = []

    context.configure(
        connection=connection,
        target_metadata=Base.metadata,
        process_revision_directives=process_revision_directives,
        transaction_per_migration=True,
        include_object=include_object,
        include_schemas=True,
        version_table_schema=settings.default_database_schema,
    )

    connection.execute(CreateSchema(settings.default_database_schema, if_not_exists=True))
    connection.execute(text(f"set search_path to {settings.default_database_schema}"))
    connection.commit()
    with context.begin_transaction():
        context.run_migrations()
```

**Comment by CaselIT** [comment]
that's definitely the issue. You either set the search path to a schema OR use a schema in the tables.
please see https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#postgresql-schema-reflection that's linked from alembic docs here https://alembic.sqlalchemy.org/en/latest/autogenerate.html#controlling-what-to-be-autogenerated

**Comment by obrazkowv** [comment]
Got it, my bad, one more time thank you for you help!
Have a good day!
